Fix: VFolder patch

Asdfx opened the pull request:
Fix patching VFolder

Commit:
Fix: VFolder patch

webinex commented on src/Webinex.Clippo/DataAccess/VRow.cs:
```diff
@@ -77,13 +77,33 @@ public static VRow<TMeta, TData> NewFile(
     }
 
-    public void UpdateFile(string name, int bytes, string @ref, string mimeType, TData data)
+    public void UpdateFile(IVFileUpdate<TData> update)
```
get back the original method and overload it with a new method. For a backward compatibility and keep it easy to use

Asdfx replied:
Changed

webinex commented on src/Webinex.Clippo/Abstractions.Extensions/VFileAddExtensions.cs:
```diff
@@ -1,23 +1,23 @@
-internal static class VFileStateExtensions
+internal static class VFileAddExtensions
 {
     public static VRow<TMeta, TData> ToRow<TMeta, TData>(
```
get back the original method and overload it with a new method. For a backward compatibility and keep it easy to use

Asdfx replied:
Why? This an internal method

webinex replied:
Okay

webinex commented on src/Webinex.Clippo.Abstractions/IVFileUpdate.cs:
```diff
@@ -0,0 +1,13 @@
+public interface IVFileUpdate<TData>
+    where TData : class, ICloneable
+{
+    public string? Name { get; }
```
I would prefer to use `Optional<T>` instead of nullability, because `null` is also a value. For example, `Name` can be nullable. Optional allows you to updated only a portion of properties.

webinex replied:
In this case `IVFilePatch` also would be nice

Asdfx replied:
Changed

webinex commented on src/Webinex.Clippo.Abstractions/IVFileAdd.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Webinex.Clippo;
+
+public interface IVFileAdd<TData>
```
`IVFileValue`?

Asdfx replied:
Renamed

webinex commented on src/Webinex.Clippo/Clippo.cs:
```diff
@@ -142,9 +142,69 @@ public async Task<IReadOnlyCollection<VFolder<TMeta, TData>>> FoldersByIdAsync(
+
+        var byFolderId = rows.GroupBy(x => x.Folder.Clone());
+
+        var result = new LinkedList<VFolder<TMeta, TData>>();
```
you can use `IAsyncEnumerable` + `yield` to make it more clear

Asdfx replied:
Changed

webinex commented on src/Webinex.Clippo/Clippo.cs:
```diff
@@ -142,9 +142,69 @@ public async Task<IReadOnlyCollection<VFolder<TMeta, TData>>> FoldersByIdAsync(
+    {
+        patches = patches.ToArray();
+
+        var byFolderId = rows.GroupBy(x => x.Folder.Clone());
```
hope there cannot be situation when patch exists, but folder not?

Asdfx replied:
There will be no error in this case as we are iterating through existing data and then applying patch to them

webinex replied:
But it might be, as for me :)

Asdfx replied:
Added error in this case

webinex commented on src/Webinex.Clippo/Clippo.cs:
```diff
@@ -142,9 +142,69 @@ public async Task<IReadOnlyCollection<VFolder<TMeta, TData>>> FoldersByIdAsync(
+        var folderRow = rows.First(r => r.Type == VRowType.Folder);
+        var fileRows = rows.Where(r => r.Type == VRowType.File).ToArray();
+
+        if (patch.Version?.HasValue == true && folderRow.Version != patch.Version.Value)
```
introduce `VFolder.ValidateVersion` method

Asdfx replied:
Introduced

webinex commented on src/Webinex.Clippo/Clippo.cs:
```diff
@@ -142,9 +142,69 @@ public async Task<IReadOnlyCollection<VFolder<TMeta, TData>>> FoldersByIdAsync(
+            return MapFolder(rows);
+        }
+
+        var newFiles = patch.Files.Where(x => x.Add != null).Select(x => x.Add!).ToArray();
```
extract this line and below to PatchFolderAndFilesAsync
AddNewFilesAsync / PatchFilesAsync / DeleteFilesAsync

Asdfx replied:
Refactored

webinex commented on src/Webinex.Clippo/Clippo.cs:
```diff
@@ -142,9 +142,69 @@ public async Task<IReadOnlyCollection<VFolder<TMeta, TData>>> FoldersByIdAsync(
+        return await PathAsync(patches, rows, meta);
+    }
+
+    private async Task<IReadOnlyCollection<VFolder<TMeta, TData>>> PathAsync(
```
typo in Pat**c**h and rename to PatchFolderAsync

Asdfx replied:
Renamed

webinex commented on src/Webinex.Clippo/Clippo.cs:
```diff
@@ -142,9 +141,80 @@ public async Task<IReadOnlyCollection<VFolder<TMeta, TData>>> FoldersByIdAsync(
+            return MapFolder(rows);
+        }
+
+        var newFileRows = await AddNewFilesAsync(patch, folderRow.Folder, meta);
```
extract to PatchFilesAsync

Asdfx replied:
Extracted

Commit:
Fix PR comments

Throw error if patch folder does not exist
Extract PatchFilesAsync from PatchFolderAsync
